packages/simcore-sdk/src/simcore_sdk/node_ports_common/filemanager.py

```diff
@@ -309,68 +310,88 @@ async def upload_file(
         progress_bar = ProgressBarData(steps=1)
 
-    use_rclone = (
+    use_rclone: bool = (
         await r_clone.is_r_clone_available(r_clone_settings)
         and store_id == SIMCORE_LOCATION
         and isinstance(file_to_upload, Path)
     )
+    num_retries: int = (
+        NodePortsSettings.create_from_envs().NODE_PORTS_IO_NUM_RETRY_ATTEMPTS
+    )
     if io_log_redirect_cb:
         await io_log_redirect_cb(f"uploading {file_to_upload}, please wait...")
     async with ClientSessionContextManager(client_session) as session:
-        upload_links = None
-        try:
-            store_id, upload_links = await get_upload_links_from_s3(
-                user_id=user_id,
-                store_name=store_name,
-                store_id=store_id,
-                s3_object=s3_object,
-                client_session=session,
-                link_type=LinkType.S3 if use_rclone else LinkType.PRESIGNED,
-                file_size=ByteSize(
-                    file_to_upload.stat().st_size
-                    if isinstance(file_to_upload, Path)
-                    else file_to_upload.file_size
-                ),
-            )
-            # NOTE: in case of S3 upload, there are no multipart uploads, so this remains empty
-            uploaded_parts: list[UploadedPart] = []
-            if use_rclone:
-                assert r_clone_settings  # nosec
-                assert isinstance(file_to_upload, Path)  # nosec
-                await r_clone.sync_local_to_s3(
-                    file_to_upload,
-                    r_clone_settings,
-                    upload_links,
-                )
-                await progress_bar.update()
-            else:
-                uploaded_parts = await upload_file_to_presigned_links(
-                    session,
-                    upload_links,
-                    file_to_upload,
-                    num_retries=NodePortsSettings.create_from_envs().NODE_PORTS_IO_NUM_RETRY_ATTEMPTS,
-                    io_log_redirect_cb=io_log_redirect_cb,
-                    progress_bar=progress_bar,
-                )
-
-            # complete the upload
-            e_tag = await _complete_upload(
-                session,
-                upload_links,
-                uploaded_parts,
-            )
-        except (r_clone.RCloneFailedError, exceptions.S3TransferError) as exc:
-            log.error("The upload failed with an unexpected error:", exc_info=True)
-            if upload_links:
-                await _abort_upload(session, upload_links, reraise_exceptions=False)
-            raise exceptions.S3TransferError from exc
-        except CancelledError:
-            if upload_links:
-                await _abort_upload(session, upload_links, reraise_exceptions=False)
-            raise
-        if io_log_redirect_cb:
-            await io_log_redirect_cb(f"upload of {file_to_upload} complete.")
-        return store_id, e_tag
+        async for attempt in AsyncRetrying(

```

Comment:
Avoid using `async for  ... with attempt` approach for retries and use instead decorators. 
It requires refactoring the "retried part" to a separate function (with a meaningful name) but it pays back since logs are **far more clear** (in the other approach there is no clear indication of **what** are you retrying)

Reply:
So I remember having a discussion with SAN about this, but I think he would not agree to have a retry at this level.
Probably somewhere else. I would wait for his input here
@sanderegg  please guide us.